src/liballoc/collections/btree/node.rs

```diff
@@ -813,20 +803,20 @@ impl<Node, Type> Handle<Node, Type> {
 }
 
 impl<BorrowType, K, V, NodeType> Handle<NodeRef<BorrowType, K, V, NodeType>, marker::KV> {
-    /// Creates a new handle to a key/value pair in `node`. `idx` must be less than `node.len()`.
-    pub fn new_kv(node: NodeRef<BorrowType, K, V, NodeType>, idx: usize) -> Self {
-        // Necessary for correctness, but in a private module
+    /// Creates a new handle to a key/value pair in `node`.
+    /// Unsafe because the caller must ensure that `idx` be less than `node.len()`.
```

Comment:
```suggestion
    /// Unsafe because the caller must ensure that `idx < node.len()`.
    /// Note that `node.len()` is always safe to call.
```

Please verify that the second line is actually true.

Reply:
I'm not sure that `node.len()` is safe. Everything I learnt over the past months tells me that it reads through a pointer that is always well aligned from memory that is always allocated, storing the value in a register or stack exclusive to the caller. But in the course of #58431, @RalfJung wrote in `into_slices_mut` that `len` wreaks havoc on things elsewhere and had to be avoided.

Reply:
Hm, it is true that under stacked borrows a call to len() can invalidate some other reference, but that doesn't really make it unsafe?

Reply:
I'm reading up on stacked borrows, and two hours later I might have understood something, i.e. what you're asking. I guess invalidating is a lot more subtle than killing. Still bad enough that it prompted the rewrite of `into_slices_mut`, but that was only to satisfy Miri and to avoid future trouble - there was no genuine UB. "Verify" means to check that Miri  is happy executing the expressions in the debug_assert (while it normally skips those).

If that's right, then yes, the second line is actually true.

Reply:
Also applied that comment style to `new_edge`

Reply:
> there was no genuine UB

I am not entirely sure what you mean by this. Stacked Borrows is experimental, so indeed nothing it says is normative, but there is general agreement that we need *something* like it  to be genuine UB.

Reply:
Ignore what I wrote (and write), I am entirely unsure what anyone means here, including myself. To me, you noticed a while ago that invoking len() in some circumstances kills a puppy somewhere, and you stopped that; Mark wrote killing puppies isn't really a bad thing, and in 68468 cuviper wrote that it's better than duplicating handles. So I figured puppies have at least as many lives as cats.

Reply:
:rofl: 

The issue with `len` is indeed that it kills ~~puppies~~pointers elsewhere, but that doesn't imply it should be `unsafe`. It just means some unsafe code should avoid (or at least be careful) calling it. A function with a `&self` argument is entirely in its right to read from all the memory reachable from `self`. It's just, some callers rely on it *not* doing that.

But that's just like a function called `return_42` that breaks code by returning 37 -- that doesn't mean the function should be `unsafe`. It is entirely normal for unsafe code to rely on more than just the UB-freedom of some other piece of safe API elsewhere. Like, all hell would break lose (and I imagine it might cause UB) if `Vec::push` was changed to be a NOP, but that doesn't change that `Vec::push` is a safe operation.

Reply:
OK, thanks, I think I understand now why the call to `len()` in `into_slices_mut` was UB.

But it doesn't bring me closer to an answer to "Please verify that the second line is actually true.", on the contrary. Now I think the statement "`Note that node.len()` is always safe to call." is false. If there were any calls in safe code, I guess it would be true but also obvious because `len()` itself is not marked unsafe. For calls in unsafe code, we have to instead list it is a precondition of `new_kv`: the caller must allow us to invoke `node.len()`.

Reply:
> For calls in unsafe code, we have to instead list it is a precondition of new_kv: the caller must allow us to invoke node.len().

Calling `len` has no preconditon other than the argument being well-typed, that's why it can be a safe function.

But it has the *effect* of killing pointers, and `new_kv` has that same effect. The possible source of trouble here is having callers that do not expect this effect. I am not sure what is the best place to document this.

Certainly, having "Note that `node.len()` is always safe to call" in the *doc comment* makes little sense here, this is justifying something internal to the function, not giving an externally visible spec.

Reply:
The intent was that the precondition is easy to assert for callers if necessary. But I don't mind much either way.

Reply:
> Calling `len` has no preconditon other than the argument being well-typed

Sure, but I was thinking of a precondition (or restriction, if precondition is the wrong term) on the unsafe `new_kv` function. Though it does seem that if you start documenting that, you'd have to do that in a lot of functions for a lot of "interactions" with memory.

Reply:
It's not a precondition, I'd say it's part of the side-effects of this function. It's part of the side-effect of *all* `&self` functions that calling it kills exclusive aliases to `self`. This one here is special only insofar as the `NodeRef` type is used instead of a "normal" reference.

Reply:
Right, I understand. "Note that `node.len()` is always safe to call" does not go into what side effects there might be.

My answer to "that the second line is actually true" is simply: yes, `NodeRef::len()` is not marked unsafe.

And Mark's point is that it is merely a hint to the caller, not a justification of the debug_assert, so there's not real reason to keep it nor to remove it.

Reply:
Yeah, let's drop it from here perhaps and make an explicit note on `NodeRef::len()`? Unfortunately given the unsafety landscape today the "is not unsafe" part of `NodeRef::len()` doesn't really actually play out to being necessarily indicative of lacking preconditions :)

Reply:
I dropped it and recycled the bytes to somewhat wishy-washy comments in `NodeRef::len()`. And not on `NodeRef::height()` or any other `&self` method...